### PR TITLE
feat(app-detailed-view): Make a GET to the API to retrieve an app by its ID (AEROGEAR-8466)

### PR DESCRIPTION
**Pull request by craicoverflow**
## Motivation

https://issues.jboss.org/browse/AEROGEAR-8466

## What

Added logic to make a GET request to the API and fetch a single app by its ID.

## Why

It is required so that we the UI can fetch live, dynamic content from the server.

## How

Added new function to `DataService.js` and updated Redux store to handle new state changes.

## Verification Steps
 
1. Make sure the API is running.
2. On the landing page, click one of the app rows to bring you to the app detailed page.
3. Open the developer console and monitor the output from the redux-logger.
4. If you see a log with action type `APP_SUCCESS` then the UI has successfully retrieved the app information from the server.
5. Navigate to the app details view by clicking on one of the apps on the landing page.
6. The table of deployed versions should render and should be sortable on all columns.

## Checklist:

- [x] Code has been tested locally by PR requester
- [ ] Changes have been successfully verified by another team member 

## Progress

- [x] Finished task

### CODE REVIEW

File: ui/src/actions/types.js [hunk 1]
```diff
@@ -1,6 +1,10 @@
 export const REVERSE_SORT = 'REVERSE_SORT';
 export const GET_APPS = 'GET_APPS';
```
Comment (laurafitzgerald) [inline ui/src/actions/types.js]: @craicoverflow am I right in saying we don't use either GET_APPS or GET_APP_BY_ID anywhere?

File: ui/src/reducers/__tests__/index.test.js [hunk 2]
```diff
@@ -1,38 +1,41 @@
...
+      { title: 'Last Launched', transforms: [sortable] },
+      { title: 'Disable on Startup', transforms: [sortable] },
+      { title: 'Custom Disable Message', transforms: [sortable] }
     ],
     isAppsRequestFailed: false,
     currentUser: 'currentUser',
-    isUserDropdownOpen: false
+    isUserDropdownOpen: false,
+    app: {
+      deployedVersions: { rows: [], data: {} }
```
Comment (laurafitzgerald) [inline ui/src/reducers/__tests__/index.test.js]: @craicoverflow there seems to already be a `deployedVersions` array inside the data object. Could we remove it from here so that the data object contains the following:


Screenshot 2019-03-14 at 17 19 33

As it stands it's nested as below:

Screenshot 2019-03-14 at 17 20 08

File: ui/src/reducers/index.js [hunk 3]
```diff
@@ -99,19 +104,43 @@ export default (state = initialState, action) => {
...
+    case APPS_FAILURE:
+      return {
+        ...state,
+        isAppsRequestFailed: true
+      };
+    case APP_REQUEST:
+      return {
+        ...state
+      };
+    case APP_SUCCESS:
```
Comment (laurafitzgerald) [inline ui/src/reducers/index.js]: I think you included this because we do the translation for the apps request in the reducer also. But because there is a React Component (the checkbox) required in the appversions table @damienomurchu did the translation in the component rather than in the reducer. We discussed moving the apps rows translation to the component again, but it's not necessary that we do that. For this case, as we're already doing it in the component we don't need to do it here.

Comment (laurafitzgerald) [reply ui/src/reducers/index.js]: In that case we probably don't even need to store the rows value in the state as we're creating that in the component and can use the json data to build it there.

Comment (craicoverflow) [reply ui/src/reducers/index.js]: Ah, I originally had it like that but changed it to keep the component changes minimal. I kept the component code mostly the same:

https://github.com/aerogear/mobile-security-service/pull/95/files#diff-2aac7a8dcccd74e488dbea5673dfc9c2R14

I will change it back.

Comment (craicoverflow) [reply ui/src/reducers/index.js]: I think we might need to store the row values in the state for sorting.

File: ui/src/containers/AppVersionTableContainer.js [hunk 4]
```diff
@@ -55,31 +52,29 @@ export class AppTableContainer extends React.Component {
...
 
 function mapStateToProps (state) {
   return {
-    app: state.appDetailRows,
-    sortBy: state.appDetailsSortDirection,
-    columns: state.appDetailColumns,
-    isAppsRequestFailed: state.isAppsRequestFailed
+    sortBy: state.appVersionsSortDirection,
+    columns: state.appVersionsColumns,
+    appVersions: state.app.deployedVersions.rows
```
Comment (damienomurchu) [inline ui/src/containers/AppVersionTableContainer.js]: Should we include also appVersions under `AppVersionTableContainer.propTypes` above?

Comment (craicoverflow) [reply ui/src/containers/AppVersionTableContainer.js]: Would that fail when appVersions is an empty array or is that valid?

File: ui/src/actions/types.js (hunk 1 above)
Comment (craicoverflow) [reply ui/src/actions/types.js]: You are right.

File: ui/src/containers/AppVersionTableContainer.js (hunk 4 above)
Comment (damienomurchu) [reply ui/src/containers/AppVersionTableContainer.js]: The propType check shouldn't - my understanding is it just checks for the presence of that prop and triggers a warning if it's not

File: ui/src/reducers/index.js (hunk 3 above)
Comment (laurafitzgerald) [reply ui/src/reducers/index.js]: @craicoverflow yep, you're 100% right. Resolving this one and there's no action required.

Comment (damienomurchu) [reply ui/src/reducers/index.js]: > I think we might need to store the row values in the state for sorting.

I think you're right. We only get the re-rendering of the table component with redux when the state mapped in `mapPropsToState` in the component changes. 

With the `AppVersionTableContainer` we need to modify the row data to create the checkbox, and we shouldn't be storing React components (like Checkbox'es) in the store, so I think we need logic in both places; in the reducer to store the row data in the state, and in the component to re-render the row data when it changes (& generate the checkboxes)

File: ui/src/containers/AppVersionTableContainer.js (hunk 4 above)
Comment (craicoverflow) [reply ui/src/containers/AppVersionTableContainer.js]: Updated.

File: ui/src/reducers/__tests__/index.test.js (hunk 2 above)
Comment (laurafitzgerald) [reply ui/src/reducers/__tests__/index.test.js]: Marking as resolved. thanks for the update

**Comment by laurafitzgerald** [comment]
I saw one small edge case where ordering on a column where all values are the same (e.g. `Last Launched` doesn't work as expected but let's resolve that in a seperate pr so that we can unblock the others. @craicoverflow would you mind creating a JIRA for that.

**Comment by craicoverflow** [comment]
> I saw one small edge case where ordering on a column where all values are the same (e.g. `Last Launched` doesn't work as expected but let's resolve that in a seperate pr so that we can unblock the others. @craicoverflow would you mind creating a JIRA for that.

Yeah I see what you mean. Will do
